extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -378,32 +406,58 @@ export class Bind {
       });
 
       const onPop = () => this.setState(oldState);
-      return this.setState(result)
-        .then(() => this.getDataForHistory_())
-        .then(data => {
-          this.history_.push(onPop, data);
+      return this.setState(result).then(() => {
+        return this.getDataForHistory_().then(data => {
+          this.historyQueue_.push({op: HistoryOp.PUSH, data, onPop});
+          // Immediately flush queue on "push" to avoid adding latency to
+          // the browser "back" functionality.
+          this.flushHistoryQueue_();
         });
+      });
     });
   }
 
+  /**
+   * @private
+   */
+  flushHistoryQueue_() {
+    for (let i = 0; i < this.historyQueue_.length; i++) {
+      const {op, data, onPop} = this.historyQueue_[i];
+      if (op === HistoryOp.REPLACE) {
+        // Skip consecutive "replace" operations.
+        const next = this.historyQueue_[i + 1];
+        if (next && next.op === HistoryOp.REPLACE) {
+          continue;
+        }
+        this.history_.replace(data);
+      } else if (op === HistoryOp.PUSH) {
+        this.history_.push(onPop, data);
+      }
+    }
+    this.historyQueue_.length = 0;
+  }
+
   /**
    * Returns data that should be saved in browser history on AMP.setState() or
    * AMP.pushState(). This enables features like restoring browser tabs.
-   * @return {!Promise<?JsonObject>}
+   * @return {!Promise<(!JsonObject|undefined)>}
    */
   getDataForHistory_() {
-    const data = dict({
-      'data': dict({'amp-bind': this.state_}),
-      'title': this.localWin_.document.title,
-    });
+    // Copy the state so subsequent changes don't modify the stored object
+    // reference, and don't pay the cost of copying JSON unless we need to.
+    const getData = () => {
+      return dict({
+        'data': dict({'amp-bind': this.copyJsonObject_(this.state_)}),
+        'title': this.localWin_.document.title,
+      });
+    };
     if (!this.viewer_.isEmbedded()) {
-      // CC doesn't recognize !JsonObject as a subtype of (JsonObject|null).
-      return /** @type {!Promise<?JsonObject>} */ (Promise.resolve(data));
+      return Promise.resolve(getData());
     }
     // Only pass state for history updates to trusted viewers, since they
     // may contain user data e.g. form input.
     return this.viewer_.isTrustedViewer().then(trusted => {
-      return trusted ? data : null;
+      return trusted ? getData() : undefined;
```

Comment:
Defering the copy like this creates a slight change of behavior, since we'll now copy the data after `isTrustedViewer` resolves instead of before.

Reply:
Yea, this is probably an unnecessary optimization since it only affects untrusted viewers. Removed.